### PR TITLE
Open component source code HopsanGUI

### PR DESCRIPTION
**Pull request by robbr48**
- Open base component sourcecode for subtype components
- Open source code from context menu
- Open source code from library widget
- Open source code from component properties dialog
- Fixed crash in restore state when text tabs are open
- Fixed wrong tab name after restore state when text tabs are open
- Prevent opening same text file in multiple tabs

### CODE REVIEW

File: HopsanGUI/Dialogs/ComponentPropertiesDialog3.cpp [hunk 1]
```diff
@@ -259,6 +260,14 @@ void ComponentPropertiesDialog3::applyPressed()
     setVariableValues();
 }
 
+void ComponentPropertiesDialog3::openSourceCode()
+{
+    auto& appearance = gpLibraryHandler->getModelObjectAppearancePtr(mpModelObject->getTypeName());
```
Comment (peterNordin) [inline HopsanGUI/Dialogs/ComponentPropertiesDialog3.cpp]: This looks strange, the function returns a shared pointer. Here a reference to the shared pointer object is created. But then the temporary sharedpointer is destroyed (I assume). Then what does the reference refere to.
I think that the & should be removed.

File: HopsanGUI/Dialogs/ComponentPropertiesDialog3.cpp [hunk 2]
```diff
@@ -259,6 +260,14 @@ void ComponentPropertiesDialog3::applyPressed()
...
+    QString basePath = appearance->getBasePath();
+    QString fileName = appearance->getSourceCodeFile();
+    gpModelHandler->loadTextFile(basePath+fileName);
```
Comment (peterNordin) [inline HopsanGUI/Dialogs/ComponentPropertiesDialog3.cpp]: I dont think that we can assume that basepath includes a trailing /, in my opinion it should not. So a / must br added, but only if basePath is not empty... we need some  clever helpfunction for this. But I assumt that the currnt  code works for some reason, maybe basepath has a trailing /

File: HopsanGUI/Dialogs/ComponentPropertiesDialog3.cpp [hunk 3]
```diff
@@ -425,6 +434,13 @@ QDialogButtonBox *ComponentPropertiesDialog3::createOKButtonBox()
     QPushButton *pOkButton = new QPushButton(tr("&Ok"), this);
     QPushButton *pApplyButton = new QPushButton(tr("&Apply"), this);
     QDialogButtonBox *pButtonBox = new QDialogButtonBox(Qt::Horizontal, this);
+    QString filePath = mpModelObject->getAppearanceData()->getSourceCodeFile();
```
Comment (peterNordin) [inline HopsanGUI/Dialogs/ComponentPropertiesDialog3.cpp]: In this case basePath is not used, maybe it should be. Or maybe getSourceCodeFile shoyld return the a solute path

File: HopsanGUI/GUIObjects/GUIModelObject.cpp [hunk 4]
```diff
@@ -1344,6 +1346,13 @@ QAction *ModelObject::buildBaseContextMenu(QMenu &rMenu, QGraphicsSceneContextMe
     {
         openPropertiesDialog();
     }
+    else if(selectedAction == sourceCodeAction)
+    {
+        auto& appearance = gpLibraryHandler->getModelObjectAppearancePtr(mModelObjectAppearance.getTypeName());
```
Comment (peterNordin) [inline HopsanGUI/GUIObjects/GUIModelObject.cpp]: Reference again, see other comment
also  below, using basePath without /, not sure if it is actually a problem

File: HopsanGUI/GUIObjects/GUIModelObjectAppearance.cpp [hunk 5]
```diff
@@ -820,6 +820,12 @@ QString ModelObjectAppearance::getBasePath() const
     return mBasePath;
 }
 
+//! @brief Get the base path that all icon paths are relative
+QFileInfo ModelObjectAppearance::getXMLFile() const
```
Comment (peterNordin) [inline HopsanGUI/GUIObjects/GUIModelObjectAppearance.cpp]: Wrong comment

File: HopsanGUI/GUIObjects/GUIModelObjectAppearance.cpp [hunk 6]
```diff
@@ -1204,6 +1210,12 @@ void ModelObjectAppearance::setBasePath(const QString path)
     setRelativePathFromAbsolute(); //Reset relative path to new basepath
 }
 
+//! @brief Access method to manually set the BaseIconPath
+void ModelObjectAppearance::setXMLFile(const QFileInfo file)
+{
+    mXMLFile = QFileInfo(file);
```
Comment (peterNordin) [inline HopsanGUI/GUIObjects/GUIModelObjectAppearance.cpp]: Wrong comment, and unnecessary? Qfileinfo copy constructor call

File: HopsanGUI/ModelHandler.cpp [hunk 7]
```diff
@@ -371,6 +371,14 @@ TextEditorWidget *ModelHandler::loadTextFile(QString fileName)
     }
     QFileInfo fileInfo(file);
 
+    //Abort if file is already open
+    for(const auto& editor : mTextEditors) {
+        if(fileInfo == editor->getFileInfo()) {
+            gpMessageHandler->addErrorMessage("File is already open: "+fileInfo.absoluteFilePath());
+            return 0;
```
Comment (peterNordin) [inline HopsanGUI/ModelHandler.cpp]: nullptr

File: HopsanGUI/Widgets/LibraryWidget.cpp [hunk 8]
```diff
@@ -864,6 +870,17 @@ void LibraryWidget::handleItemClick(QTreeWidgetItem *item, int column)
             {
                 QDesktopServices::openUrl(QUrl("file:///" + gpLibraryHandler->getModelObjectAppearancePtr(mItemToTypeNameMap.find(pFirstSubComponentItem).value())->getBasePath()));
             }
+            else if(pReply == pEditXMLAction) {
+                QFileInfo xmlFile = gpLibraryHandler->getModelObjectAppearancePtr(mItemToTypeNameMap.find(pFirstSubComponentItem).value())->getXMLFile();
+                gpModelHandler->loadTextFile(xmlFile.absoluteFilePath());
+            }
+            else if(pReply == pEditCodeAction) {
+                QString typeName = gpLibraryHandler->getModelObjectAppearancePtr(mItemToTypeNameMap.find(pFirstSubComponentItem).value())->getTypeName();
+                auto& appearance = gpLibraryHandler->getModelObjectAppearancePtr(typeName);
```
Comment (peterNordin) [inline HopsanGUI/Widgets/LibraryWidget.cpp]: Reference and assuming basePath has trailing slash (see other comments)

File: HopsanGUI/Dialogs/ComponentPropertiesDialog3.cpp (hunk 1 above)
Comment (robbr48) [reply HopsanGUI/Dialogs/ComponentPropertiesDialog3.cpp]: True. I will remove the &.

File: HopsanGUI/Dialogs/ComponentPropertiesDialog3.cpp (hunk 2 above)
Comment (robbr48) [reply HopsanGUI/Dialogs/ComponentPropertiesDialog3.cpp]: Perhaps all functions returning a path should be responsible of adding a trailing slash if it is missing? I will just add an if statement to solve it for now.

File: HopsanGUI/Dialogs/ComponentPropertiesDialog3.cpp (hunk 3 above)
Comment (robbr48) [reply HopsanGUI/Dialogs/ComponentPropertiesDialog3.cpp]: This variable is only used to check if there is a source file specified, to determine if the button should be visible or not. A bit stupid I guess, but the alternative would be to add a "hasSourceCodeFile()" function, which is not much better.

File: HopsanGUI/Dialogs/ComponentPropertiesDialog3.cpp (hunk 2 above)
Comment (peterNordin) [reply HopsanGUI/Dialogs/ComponentPropertiesDialog3.cpp]: No, this is how it was previously solved, but then sometimes you do not want a trailing slash, I have had lots of problems with this sort of thing. The correct way is to never include a trailing slash, and then use a helpfunction similar to pythons path.join(part1,part2) that is clever and adds / as needed. But we can add that later

**Pull request merged**